File: python/algo/202501/algo0969.py
from typing import List

# 贪心专题 - 构造题
class Solution:
    def pancakeSort(self, arr: List[int]) -> List[int]:   
        def get_index(nums: List[int], val:int) -> int:
            for i, v in enumerate(nums):
                if v == val: return i
        
        ans = []
        while len(arr) > 1:
            n = len(arr)
            inx = get_index(arr, n)
            if inx + 1 != n:
                ans.extend([inx+1, n])
                arr = arr[inx+1:][::-1] + arr[:inx]
            else:
                arr = arr[:inx]
        return ans

File: cplus/leetcode/202501/algo0969.cpp
#include <bits/stdc++.h>
using namespace std;

// 贪心专题 - 构造题
class Solution {
public:
    vector<int> pancakeSort(vector<int>& arr) {
        // 参考题解，学习 c++ 语法，其实也特别方便
        vector<int> ans;

        for (int i=arr.size(); i; i--) {
            int inx = max_element(arr.begin(), arr.begin()+i) - arr.begin();
            if (inx + 1 == i) continue;

            reverse(arr.begin(), arr.begin() + inx + 1);
            reverse(arr.begin(), arr.begin() + i);
            ans.push_back(inx+1);
            ans.push_back(i);
        }

        return ans;
    }
};